coderd/prometheusmetrics/aggregator.go

```diff
@@ -0,0 +1,249 @@
+	updateHistogram := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "coderd",
+		Subsystem: "prometheusmetrics",
+		Name:      "metrics_aggregator_execution_update_seconds",
+		Help:      "Histogram for duration of metrics aggregator update in seconds.",
+		Buckets:   []float64{0.001, 0.005, 0.010, 0.025, 0.050, 0.100, 0.500, 1, 5, 10, 30},
+	})
+	err := registerer.Register(updateHistogram)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanupHistogram := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "coderd",
+		Subsystem: "prometheusmetrics",
+		Name:      "metrics_aggregator_execution_cleanup_seconds",
+		Help:      "Histogram for duration of metrics aggregator cleanup in seconds.",
+		Buckets:   []float64{0.001, 0.005, 0.010, 0.025, 0.050, 0.100, 0.500, 1, 5, 10, 30},
+	})
+	err = registerer.Register(cleanupHistogram)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricsAggregator{
+		log:                    logger,
+		metricsCleanupInterval: metricsCleanupInterval,
+
+		collectCh: make(chan (chan<- prometheus.Metric), sizeCollectCh),
+		updateCh:  make(chan updateRequest, sizeUpdateCh),
+
+		updateHistogram:  updateHistogram,
+		cleanupHistogram: cleanupHistogram,
+	}, nil
+}
+
+func (ma *MetricsAggregator) Run(ctx context.Context) func() {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	done := make(chan struct{})
+
+	cleanupTicker := time.NewTicker(ma.metricsCleanupInterval)
+	go func() {
+		defer close(done)
+		defer cleanupTicker.Stop()
+
+		for {
+			select {
+			case req := <-ma.updateCh:
+				ma.log.Debug(ctx, "metrics aggregator: update metrics")
+
+				timer := prometheus.NewTimer(ma.updateHistogram)
+			UpdateLoop:
+				for _, m := range req.metrics {
+					for i, q := range ma.queue {
+						if q.username == req.username && q.workspaceName == req.workspaceName && q.agentName == req.agentName && q.Name == m.Name {
+							ma.queue[i].AgentMetric.Value = m.Value
+							ma.queue[i].expiryDate = req.timestamp.Add(ma.metricsCleanupInterval)
+							continue UpdateLoop
+						}
+					}
+
+					ma.queue = append(ma.queue, annotatedMetric{
+						username:      req.username,
+						workspaceName: req.workspaceName,
+						agentName:     req.agentName,
+
+						AgentMetric: m,
+
+						expiryDate: req.timestamp.Add(ma.metricsCleanupInterval),
+					})
+				}
+
+				timer.ObserveDuration()
+			case inputCh := <-ma.collectCh:
+				ma.log.Debug(ctx, "metrics aggregator: collect metrics")
+
+				for _, m := range ma.queue {
+					desc := prometheus.NewDesc(m.Name, metricHelpForAgent, agentMetricsLabels, nil)
+					valueType, err := asPrometheusValueType(m.Type)
+					if err != nil {
+						ma.log.Error(ctx, "can't convert Prometheus value type", slog.F("name", m.Name), slog.F("type", m.Type), slog.F("value", m.Value), slog.Error(err))
+						continue
+					}
+					constMetric := prometheus.MustNewConstMetric(desc, valueType, m.Value, m.username, m.workspaceName, m.agentName)
+					inputCh <- constMetric
```

Comment:
I'm not sure how this will be used, but I worry about the use of channels. Essentially the way this behaves now is that it's enough for one misbehaving consumer to lock up this entire goroutine since it's dependent on the collected metrics being consumed in a timely manner.

So essentially we can insert 128 entries here immediately, but if more is queued up then it's on the one who called `Collect(ch)` to ensure things progress smoothly. This could prevent update and cleanup.

Reply:
> So essentially we can insert 128 entries here immediately, but if more is queued up then it's on the one who called Collect(ch) to ensure things progress smoothly. This could prevent update and cleanup.

128 entries correspond to 128 agents submitting their metrics at the same time. I'm aware of the issue, but considering our scale, I don't think that we should encounter this problem soon. Speaking of a potential solution, do you think that we should introduce a "middle channel" to timeout the `Collect(ch)` if metrics are not streamed? I'm afraid that it may complicate the existing logic.

Reply:
Hmm, I see. I think another option, instead of an intermediate channel would be to send a slice over this channel instead of 128 items over the channel. Then Collect can process the slice instead of consuming the channel and the goroutine can immediately continue its work. This way the logic doesn't need to change much at all. This ofc leads to increased memory usage when we exceed 128, but then again, if we do exceed it now it might not be ideal either.

Reply:
> I think another option, instead of an intermediate channel would be to send a slice over this channel instead of 128 items over the channel. Then Collect can process the slice instead of consuming the channel and the goroutine can immediately continue its work. 

In this case `Collect` will need to know when the slice is ready to be "ranged over", otherwise it would be a data race.

I slightly modified your idea. I kept the channel, but I'm passing the entire slice at once, so 128 items limit does not exist anymore. Please let me know your thoughts, @mafredri.

Reply:
Nice, and it actually looks like how I intended it 😄